### PR TITLE
PackageReference support for .NET Core C++/CLI projects

### PR DESCRIPTION
**Pull request by nkolev92**
Design for https://github.com/NuGet/Home/issues/10194. 

cc @olgaark @davkean @zivkan 

[Rendered](https://github.com/NuGet/Home/blob/dev-nkolev92-cpluspluscli/proposed/2021/NuGet-CPPCLI-Support.md)

### CODE REVIEW

File: designs/NuGet-CPPCLI-Support.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+TargetFramework*/TargetPlatform* properties for C++/CLI are pretty much the same as for c#, so they should be used in nuget instead of just checking for .vcxproj extension. Also, C++/CLI can use/reference not only managed, but native code as well, so “native” packages should be allowed together with the appropriate managed ones.
+You can use the following properties to distinguish vc projects:
+TargetPlatformIdentifier/Version/Moniker should be defined for all vc projects.
+CLRSupport
+False or empty – native project
+NetCore – C++/CLI .NET Core project
+Everything else (true, safe, etc)  - C++/CLI .NET Framework project
+TargetFramework is only defined for C++/CLI .NET Core projects.
+TargetFrameworkIdentifier/Version/Moniker properties are defined for C++/CLI .NET Framework projects. Note, that they are also confusingly defined for pure native ones too (in msbuild common targets, at least, they used to be), so you need to check other properties first (i.e. TargetPlatformIdentifier and CLRSupport). You can also use project capabilities: “native” vs “managed”
+```
```
Comment (zivkan) [inline designs/NuGet-CPPCLI-Support.md]: This section is difficult to understand. In fact, I only understand about half of what I think it's trying to communicate to me.

File: designs/NuGet-CPPCLI-Support.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+Customers using C++/CLI projects.
+
+## Goals
+
+Support the C++ team as necessary to provide NuGet PackageReference support for C++/CLI projects.
+
+At this point focused to:
+
+* Ensure NuGet is interpreting the correct framework.
```
Comment (zivkan) [inline designs/NuGet-CPPCLI-Support.md]: CLI only, or VS as well?

I assume in VS, the C++ project system will be expected to use the CPS Package Reference "contract", meaning they'll send project nominations as required?

File: designs/NuGet-CPPCLI-Support.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+Currently the steps for NuGet framework inference for project types still actively supported are as follows:
+
+1. Does the project file extension end with `.vcxproj`? Then `native0.0`
+1. Does `TargetPlatformMoniker` start with `UAP`? Then the framework is `UAP{version}`, where `TargetPlatformMinVersion` is preferred over `TargetPlatformVersion`.
+1. `TargetPlatformMoniker`, unless the framework is `.NETCoreApp` with version greater than 5.0, then it's the combination of `TargetFrameworkMoniker` and `TargetPlatformMoniker`.
+
+To understand the proposed changes, here's a mapping of all `C++` project types and the property value for the appropriate ones.
+
+| Project type | TargetFrameworkIdentifier | TargetPlatformIdentifier | CLRSupport | Effective NuGet framework | Notes | Open questions |
```
Comment (zivkan) [inline designs/NuGet-CPPCLI-Support.md]:
```suggestion
| Project type | TargetFrameworkMoniker | TargetPlatformMoniker | CLRSupport | Effective NuGet framework | Notes | Open questions |

```

File: designs/NuGet-CPPCLI-Support.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+|--------------|---------------------------|--------------------------|------------|---------------------------|-------|----------------|
+| Native C++ | .NETFramework,Version=v4.0 | Windows,Version=10.0.19041.0 | false | native | NuGet will continue special casing vcxproj. |
```
Comment (zivkan) [inline designs/NuGet-CPPCLI-Support.md]: native c++ identifies as .NET Framework 4.0?

File: designs/NuGet-CPPCLI-Support.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+Similar to how `.NET Core` implements `.NET Standard` and supports `.NET Core` as a same family framework, we can add a framework type that's only allowed in projects, and not packages. This framework could define it's compatible as both `native` and `managed`.
+
+The downside here is that NuGet does not have a framework that's only supported in projects. Given that this is not a real `target`, it is counter to what NuGet frameworks are supposed to represent.
+
+## Test Strategy
+
+TBD
+
+## Future Work
```
Comment (zivkan) [inline designs/NuGet-CPPCLI-Support.md]: APIs for how NuGet is supposed to inform the C++ project system about modifications on package install/uninstall/upgrade?

File: designs/NuGet-CPPCLI-Support.md (hunk 1 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: I didn't write it. It's a verbatim answer from the e-mail thread to the question of how should NuGet detect the framework.

File: designs/NuGet-CPPCLI-Support.md (hunk 2 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: commandline and VS. 
The C++/CLI projects already nominate today (it was work done in 16.9), but we don't use the correct framework so you can't install managed packages.

File: designs/NuGet-CPPCLI-Support.md (hunk 4 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: Native C++ has a default. 
The `CLRSupport` flag will be the first thing read, `false` or `` means native.

File: designs/NuGet-CPPCLI-Support.md (hunk 5 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: You mean Package Manager UI changes?
That'll go through `UnconfiguredProject`, same for managed/SDK-based CPS PR projects.

File: designs/NuGet-CPPCLI-Support.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
+# NuGet C++/CLI support
+
+* Status: In Review
+* Author(s): [Nikolche](https://github.com/nkolev92)
+* Issue: [8195](https://github.com/NuGet/Home/issues/8195) and [10194](https://github.com/NuGet/Home/issues/10194) NuGet support for C++/CLI projects.
+
+## Problem Background
+
+C++/CLI is a project type that allows C++ programs to use .NET classes.
```
Comment (kartheekp-ms) [inline designs/NuGet-CPPCLI-Support.md]: I see following templates available for language `C++` and project type  `Console` in VS 2019. Is this spec appliacable for all the following project tempates?

image

File: designs/NuGet-CPPCLI-Support.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+Currently the steps for NuGet framework inference for project types still actively supported are as follows:
+
+1. Does the project file extension end with `.vcxproj`? Then `native0.0`
+1. Does `TargetPlatformMoniker` start with `UAP`? Then the framework is `UAP{version}`, where `TargetPlatformMinVersion` is preferred over `TargetPlatformVersion`.
+1. `TargetPlatformMoniker`, unless the framework is `.NETCoreApp` with version greater than 5.0, then it's the combination of `TargetFrameworkMoniker` and `TargetPlatformMoniker`.
+
+To understand the proposed changes, here's a mapping of all `C++` project types and the property value for the appropriate ones.
+
+| Project type | TargetFrameworkMoniker | TargetPlatformMoniker | CLRSupport | Effective NuGet framework | Notes | Open questions |
```
Comment (kartheekp-ms) [inline designs/NuGet-CPPCLI-Support.md]: This table is not rendering properly in the [markdown](https://github.com/NuGet/Home/blob/dev-nkolev92-cpluspluscli/designs/NuGet-CPPCLI-Support.md#solution-overview)

File: designs/NuGet-CPPCLI-Support.md (hunk 2 above)
Comment (kartheekp-ms) [reply designs/NuGet-CPPCLI-Support.md]: Are we referring to `dotnet CLI` while referring `commandline` in the above conversation?

Can `C++/CLI projects` use `dotnet restore` after we implement this change as they target `.NET Core`?

File: designs/NuGet-CPPCLI-Support.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+## Non-Goals
+
+* packages.config support for C++/CLI projects. Focus is on .NET Core scenarios right now.
+
+## Solution Overview
+
+In managed projects, NuGet interprets the effective framework by looking at `TargetFrameworkMoniker`, `TargetPlatformMoniker` and in UWP cases, `TargetPlatformMinVersion`.
+
+C++ projects have an additional property CLRSupport that's set for C++ projects.
```
Comment (kartheekp-ms) [inline designs/NuGet-CPPCLI-Support.md]: As per [blog](https://devblogs.microsoft.com/cppblog/porting-a-c-cli-project-to-net-core/#migrating-a-vcxproj-to-net-core), `CLRSupport` property is set to `NetCore` and `TargetFramework` set to specific `netcore tfm`. For e.g., 
```xml
<CLRSupport>NetCore</CLRSupport>
<TargetFramework>netcoreapp3.1</TargetFramework>
```
I think we need to use both `CLRSupport` and `TargetFramework` properties to enable package reference. Thoughts?

File: designs/NuGet-CPPCLI-Support.md (hunk 6 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: No, that's not it. Those are C++ only.

Search for C++/CLI. 
The templates will say something, Provides interopability between .NET and C++ Code.

File: designs/NuGet-CPPCLI-Support.md (hunk 8 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: We don't, `TargetFramework` is not a property that NuGet parses. 
For framework inference NuGet uses `TargetFrameworkMoniker` and `TargetPlatformMoniker`. 

NuGet used to use `TargetFramework` directly, but that was a problem. Now the SDK handles that.

File: designs/NuGet-CPPCLI-Support.md (hunk 2 above)
Comment (olgaark) [reply designs/NuGet-CPPCLI-Support.md]: [C++/CLI ](https://en.wikipedia.org/wiki/C%2B%2B/CLI#:~:text=%28March%202015%29%20C%2B%2B%2FCLI%20%28C%2B%2B%20modified%20for%20Common%20Language,interoperability%20with%20Microsoft%20.Net%20languages%20such%20as%20C%23.) is a language extension and has nothing to do with dotnet CLI. 
dotnet.exe does not currently support C++/CLI vc projects.

File: designs/NuGet-CPPCLI-Support.md (hunk 7 above)
Comment (nkolev92) [reply designs/NuGet-CPPCLI-Support.md]: Thanks...worked at some point, lol. 

Fixed now.

**Comment by JonDouglas** [comment]
Hello!

I'm working to cleanup stale PRs on `NuGet/Home`. 

Actions to take:

- If this PR **should be merged** into `NuGet/Home`, please assign a handful of reviewers or ask for another review to approve a merge.
- If this PR **should be abandoned**, please close it.
- If this PR **should remain open**, please bump the conversation forward or leave a comment for this to not be considered abandoned in a future pass.

If there's anything I can do to help out, please let me know!

File: proposed/2021/NuGet-CPPCLI-Support.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+To understand the proposed changes, here's a mapping of all `C++` project types and the property value for the appropriate ones.
+
+| Project type | TargetFrameworkIdentifier | TargetPlatformIdentifier | CLRSupport | Effective NuGet framework | Notes | Open questions |
+|--------------|---------------------------|--------------------------|------------|---------------------------|-------|----------------|
+| Native C++ | .NETFramework,Version=v4.0 | Windows,Version=10.0.19041.0 | false | native | NuGet will continue special casing vcxproj. |
+| C++ UWP App | | UAP,Version=10.0.18362.0 | false | native | NuGet will continue special casing vcxproj. |
+| CLR C++ | .NETFramework,Version=v4.7.2 | Windows,Version=10.0.19041.0 | true | ??? | This project has .NET Framework CLR support | Are these projects expected to be PackageReference or packages.config? What's the effective target framework |
+| Core CLR C++ | .NETCoreApp,Version=v5.0 | Windows,Version=10.0.19041.0 | NetCore | ??? | What's the effective framework? Is it `net5.0-windows`? |
+
```
Comment (zkat) [inline proposed/2021/NuGet-CPPCLI-Support.md:95]: I'm confused a bit here: only `.NETCoreApp,Version=v5.0.0` and up should have a TPI. Other types don't support platforms at all.

Comment (nkolev92) [reply proposed/2021/NuGet-CPPCLI-Support.md:95]: There are defaults set in the projects. 
I'm currently updating the spec, I'll take a look whether it needs clarification.

**Comment by nkolev92** [comment]
There's still some open questions we are resolving @NuGet/nuget-client team. 

So a team review is not necessarily required unless you are curious about the progress :)

**Comment by nkolev92** [comment]
Hey all, 

Can someone take a look and approve this design so I can merge?

The implementation in https://github.com/NuGet/NuGet.Client/pull/4076 follows this exact design.

**Review by zkat — APPROVED** [review]

**Review by zivkan — APPROVED** [review]

**Pull request merged**